src/main/java/org/eclipse/uprotocol/communication/InMemorySubscriber.java

```diff
@@ -112,22 +158,19 @@ public CompletionStage<UStatus> unsubscribe(UUri topic, UListener listener, Call
         Objects.requireNonNull(topic, "Unsubscribe topic missing");
         Objects.requireNonNull(listener, "listener missing");
 
-        final UUri unsubscribe = UriFactory.fromProto(
-            USubscriptionProto.getDescriptor().getServices().get(0), METHOD_UNSUBSCRIBE);
         final UnsubscribeRequest unsubscribeRequest = UnsubscribeRequest.newBuilder().setTopic(topic).build();
         
-        return RpcMapper.mapResponseToResult(rpcClient.invokeMethod(
-            unsubscribe, UPayload.pack(unsubscribeRequest), options), UnsubscribeResponse.class)
-            .thenApplyAsync(response -> {
+        return RpcMapper.mapResponseToResult(
+            // Send the unsubscribe request
+            rpcClient.invokeMethod(UNSUBSCRIBE_METHOD, UPayload.pack(unsubscribeRequest), options), 
+            UnsubscribeResponse.class)
+            
+            // Then unregister the listener
+            .thenCompose(response -> {
                 if (response.isSuccess()) {
-                    transport.unregisterListener(topic, listener).thenAccept(status -> {
-                        if (status.getCode() != UCode.OK) {
-                            throw new UStatusException(status);
-                        }
-                    });
-                    return UStatus.newBuilder().setCode(UCode.OK).build();
+                    return transport.unregisterListener(topic, listener);
```

Comment:
If this fails, we end up in an inconsistent state, don't we? We are (logically) unsubscribed but still have the listener registered ...

Reply:
Very good point, I will fix this thanks @sophokles73